File: app/src/main/java/knf/kuma/videoservers/OkruServer.java
package knf.kuma.videoservers;

import android.content.Context;
import android.support.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;
import org.jsoup.Jsoup;

import knf.kuma.commons.BypassUtil;
import knf.kuma.commons.PatternUtil;

import static knf.kuma.videoservers.VideoServer.Names.OKRU;

public class OkruServer extends Server {
    public OkruServer(Context context, String baseLink) {
        super(context, baseLink);
    }

    @Override
    public boolean isValid() {
        return baseLink.contains("server=ok");
    }

    @Override
    public String getName() {
        return OKRU;
    }

    @Nullable
    @Override
    public VideoServer getVideoServer() {
        try {
            String frame = baseLink.substring(baseLink.indexOf("'") + 1, baseLink.lastIndexOf("'"));
            String down_link = Jsoup.parse(frame).select("iframe").first().attr("src");
            String true_link = PatternUtil.extractOkruLink(Jsoup.connect(down_link).cookies(BypassUtil.getMapCookie(context)).userAgent(BypassUtil.userAgent).get().select("script").last().html());
            String e_json = Jsoup.connect(true_link).get().select("div[data-module='OKVideo']").first().attr("data-options");
            String cut_json = "{" + e_json.substring(e_json.lastIndexOf("\\\"videos"), e_json.indexOf(",\\\"metadataEmbedded")).replace("\\&quot;", "\"").replace("\\u0026", "&").replace("\\", "").replace("%3B", ";") + "}";
            JSONArray array = new JSONObject(cut_json).getJSONArray("videos");
            VideoServer videoServer = new VideoServer(OKRU);
            for (int i = 0; i < array.length(); i++) {
                JSONObject object = array.getJSONObject(i);
                switch (object.getString("name")) {
                    case "hd":
                        videoServer.addOption(new Option(getName(), "HD", object.getString("url")));
                        break;
                    case "sd":
                        videoServer.addOption(new Option(getName(), "SD", object.getString("url")));
                        break;
                    case "low":
                        videoServer.addOption(new Option(getName(), "LOW", object.getString("url")));
                        break;
                    case "lowest":
                        videoServer.addOption(new Option(getName(), "LOWEST", object.getString("url")));
                        break;
                    case "mobile":
                        videoServer.addOption(new Option(getName(), "MOBILE", object.getString("url")));
                        break;
                }
            }
            return videoServer;
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/knf/kuma/videoservers/OkruServer.kt
package knf.kuma.videoservers

import android.content.Context
import knf.kuma.commons.BypassUtil
import knf.kuma.commons.PatternUtil
import knf.kuma.commons.iterator
import knf.kuma.uagen.randomLatestUA
import knf.kuma.uagen.randomWindowsUA
import knf.kuma.videoservers.VideoServer.Names.OKRU
import kotlinx.coroutines.runBlocking
import org.apache.commons.text.StringEscapeUtils
import org.json.JSONObject
import org.jsoup.Jsoup
import org.jsoup.helper.HttpConnection

class OkruServer(context: Context, baseLink: String) : Server(context, baseLink) {

    override val isValid: Boolean
        get() = baseLink.contains("ok.ru")

    override val name: String
        get() = OKRU

    override val videoServer: VideoServer?
        get() {
            try {
                val downLink = PatternUtil.extractLink(baseLink)
                val (source) = Regex("data-options=\"(.*?)\"").find(Jsoup.connect(downLink).get().html())?.destructured!!
                val jsonText = StringEscapeUtils.unescapeHtml4(source)
                val videos = JSONObject(JSONObject(jsonText).getJSONObject("flashvars").getString("metadata")).getJSONArray("videos")
                val options = mutableListOf<Option>()
                for(video in videos){
                    val url = video.getString("url")
                    val name = when(video.getString("name")){
                        "mobile" -> "144p"
                        "lowest" -> "240p"
                        "low" -> "360p"
                        "sd" -> "480p"
                        "hd" -> "720p"
                        "full" -> "1080p"
                        "quad" -> "2000p"
                        "ultra" -> "4000p"
                        else -> "Default"
                    }
                    options.add(Option(this.name, name, url, Headers().apply { addHeader("User-Agent", HttpConnection.DEFAULT_UA) }))//mapOf("User-agent" to userAgent)
                }
                check(options.isNotEmpty())
                return VideoServer(OKRU, options, true)
            } catch (e: Exception) {
                e.printStackTrace()
                return null
            }
        }
}
